### ISSUE TITLE
[no-unused-vars]: reporter should report the last reference instead of declaration

### ISSUE THREAD

**Issue by anikethsaha** [eslint/eslint#13181]
**What rule do you want to change?**
`no-unused-vars`

**Does this change cause the rule to produce more or fewer warnings?**
same

**How will the change be implemented? (New option, new default behavior, etc.)?**
default

**Please provide some example code that this change will affect:**
Currently, the reporter reports the location in a declaration of a variable which is fine but not for all cases. 

`case 1`
```js
const unusedVar = 1; // reporting here is ok
```

`case2`
```js
let a = 1 // currently its reporting here,  reporting here is not ok as `a` has references though its being self referred 
a = a + 1;
a = a + 2;
a = a + 3; // it should report here
```

In this way, it makes easy to track the error if any.
Another advantage of reporting this way is many editors do provide shortcut keys to jump to the declaration of any variable/function but the opposite is not true for some. 

ref : https://github.com/eslint/eslint/issues/12871, https://github.com/eslint/eslint/issues/13172

**What does the rule currently do for this code?**
reports in the variable declaration

**What will the rule do after it's changed?**
reports in the last reference of the variable if the variable is being self-referred

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by nzakas** [comment]
Thanks for writing this up. Just to make sure I understand, you’re suggesting that the last reference of an unused variable be highlighted? And if there is only a declaration then the declaration is highlighted?

**Comment by anikethsaha** [comment]
Yes

**Comment by nzakas** [comment]
That sounds good to me. It would be nice if we could use a clearer message to, something like "The value of variable foo is never used after this assignment" or something similar.

**Comment by anikethsaha** [comment]
> "The value of variable foo is never used after this assignment"

much better 👍

**Comment by Lonniebiz** [comment]
This bug confused me (as shown in #13227).

**Comment by anikethsaha** [comment]
I have submitted the PR https://github.com/eslint/eslint/pull/13354 to move this issue forward quickly.

**Comment by Lonniebiz** [comment]
@nzakas Thanks. Hey, got any new books for me to buy? I always buy your books. I can proof-read too.
